ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -0,0 +1,557 @@
+isOpenDBImpl HasFS{..} _err VolatileDBEnv{..} = do
+    mSt <- atomically (readTMVar _dbInternalState)
+    return $ isJust mSt
+
+reOpenDBImpl :: (HasCallStack, MonadMask m, MonadSTM m, Ord blockId, Show blockId)
+             => HasFS m
+             -> ErrorHandling (VolatileDBError blockId) m
+             -> VolatileDBEnv m blockId
+             -> m ()
+reOpenDBImpl hasFS@HasFS{..} err VolatileDBEnv{..} = do
+    modifyTMVar _dbInternalState $ \mbSt -> case mbSt of
+        Just (st@InternalState{..}) -> return (Just st, ())
+        Nothing -> do
+            st <- mkInternalStateDB hasFS err _dbFolder _parser _maxBlocksPerFile _toSlot
+            return (Just st, ())
+
+getBlockImpl :: (MonadSTM m, MonadMask m, Ord blockId, Show blockId)
+             => HasFS m
+             -> ErrorHandling (VolatileDBError blockId) m
+             -> VolatileDBEnv m blockId
+             -> blockId
+             -> m (Maybe ByteString)
+getBlockImpl hasFS@HasFS{..} err env@VolatileDBEnv{..} slot = do
+    modifyState hasFS err env $ \st@InternalState{..} -> do
+        case Map.lookup slot _currentRevMap of
+            Nothing -> return (st, Nothing)
+            Just (file, w, n) ->  do
+                bs <- withFile hasFS (_dbFolder ++ [file]) IO.ReadMode $ \hndl -> do
+                        _ <- hSeek hndl IO.AbsoluteSeek w
+                        hGet hndl n
+                return (st, Just bs)
+
+-- This function follows the approach:
+-- (1) hPut bytes to the file
+-- (2) if full hClose the write file
+-- (3)         hOpen a new write file
+-- (4) update the Internal State.
+-- If there is an error after (1) or after (2) we should make sure that when we reopen a db from scratch,
+-- it can succesfully recover if it does not find an empty file to write and all other files are full.
+-- We should also make sure that the fs can be recovered if we get an exception/error at any moment
+-- and that we are left with an empty Internal State.
+-- We should be careful about not leaking open fds when we open a new file, since this can affect garbage
+-- collection of files.
+putBlockImpl :: forall m. (MonadMask m, MonadSTM m)
+             => forall blockId. (Ord blockId, Show blockId)
+             => HasFS m
+             -> ErrorHandling (VolatileDBError blockId) m
+             -> (blockId -> Slot)
+             -> VolatileDBEnv m blockId
+             -> blockId
+             -> BS.Builder
+             -> m ()
+putBlockImpl hasFS@HasFS{..} err toSlot env@VolatileDBEnv{..} blockId builder = do
+    modifyState hasFS err env $ \st@InternalState{..} -> do
+        case Map.lookup blockId _currentRevMap of
+            Just _ -> return (st, ()) -- trying to put an existing blockId is a no-op.
+            Nothing -> do
+                (mmaxSlot, nBlocks, fileMp) <-
+                    case Map.lookup _currentWritePath _currentMap of
+                        Nothing -> EH.throwError err $ UndisputableLookupError _currentWritePath _currentMap
```

Comment:
Can add a TODO here also to make this an internal function (and use `error`).

Reply:
@edsko You mean `error` instead of `EH.throwError`?

Reply:
We discussed this in the call no? This is an internal invariant violation, it's not an error that should ever happen? Am I confusing it with something else?

Reply:
Yes that's true. I just did't remember that on internal invariant violations we should not follow the same approach as other errors. I actually have some tests with this error: I intentionally corrupt the internal state and check if this error is indeed raised. Maybe `error` would make these tests a bit weird to write?

Reply:
I think you can just delete those tests. Invariants are there for a reason; seeing how the code behaves when the invariants are broken isn't very useful, we can't say all that much meaningful about it.

If you are worried about the invariant being broken, a better test is to check for that instead: no matter what pattern of calls into the API we have, the internal invariant is always satisfied. So, write the invariant as a function, and check it. That _would_ be a very useful test.